designs/node-overlays.md

```diff
@@ -0,0 +1,186 @@
+
+A [previous proposal](https://github.com/aws/karpenter-provider-aws/pull/2390/files#diff-b7121b2d822e57b70203794f3b6367e7173b16d84070c3712cd2f15b8dbd11b2R133) explored the tradeoffs between a 1:1 semantic vs more flexible alternatives. It raised valid observability challenges around communicating to customers exactly how the CR was modifying the scheduling simulation. It recommended a 1:1 approach, with the assumption that some system would likely dynamically codegenerate the CRs for the relevant use case. We received [direct feedback](https://github.com/aws/karpenter-provider-aws/pull/2404#issuecomment-1250688882) from some customers that the number of variations would be combinatorial, and thus unmanagable for their use case. Ultimately, we paused this analysis in favor of other priorities.
+
+After two years of additional customer feedback in this problem space, this proposal asserts that the flexibility of a selector-based approach is a signficant simplification for customers compared to a 1:1 approach and avoids the scalability challenges associated with persisting large numbers of CRs in the Kubernetes API. This decision will require us to work through the associated observability challenges.
+
+### Merging
+
+Our decision to pursue Selectors that allow for an n:m semantic presents a challenge where multiple NodeOverlays could match a simulation, but with different values. To address this, we will support an optional `.spec.weight` parameter (discussed in the [previous proposal](https://github.com/aws/karpenter-provider-aws/pull/2390/files#diff-b7121b2d822e57b70203794f3b6367e7173b16d84070c3712cd2f15b8dbd11b2R153)).
+
+We have a few choices for dealing with conflicting NodeOverlays
+
+1. Multiple NodeOverlays are merged in order of weight
+2. Multiple NodeOverlays are applied sequentially in order of weight
+3. One NodeOverlay is selected by weight
+
+We recommend option 1, as it supports the common use case of default + override semantic. In the memory overhead example, both NodeOverlays attempt to define the same parameter, but one is more specific than the other. Option 1 and 3 allow us to support this use case, where option 2 would result in 50Mi+10Mi. We are not aware of use cases that would benefit from sequential application.
+
+We choose option 1 over option 3 to support NodeOverlays that define multiple values. For example, we could collapse the example NodeOverlays using a merge semantic.
+
+```yaml
+kind: NodeOverlay
+metadata:
+  name: default
+spec:
+  pricePercent: 90
+  overhead:
+    memory: 10Mi
```

Comment:
My guess is that price discounts vary by a different dimension than by how overhead will vary. Overhead could vary by instance size, especially if modeled as a percent. If it's flat, it may be properly modeled like this.

Reply:
Yeah strongly agree. The merge semantic is a solve for this.